### ISSUE TITLE
Allocation of non-zeroed memory for performance boost

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#14687]
The default malloc implementation `GC.malloc` is supposed to zero the allocated memory. This is a sane default choice because it ensures a clean starting field. The GC won't risk incorrectly assuming uninitialized memory to contain live pointers.

But zeroing out memory has a runtime cost. And it's often useless because the memory gets fully initialized and all zeroed bits are overridden right away (e.g: https://github.com/crystal-lang/crystal/pull/14671#discussion_r1632103195).

Performance could be improved if there was a way to avoid zeroing when not necessary.

A potential implementation is proposed in #14679

**Comment by ysbaddaden** [comment]
As stated in #14679: I'd like to avoid introducing `calloc(size)` in the public API when the C function is `calloc(num, size)`. I understand we don't care about the `num` arg, but that's introducing a discrepancy over a weighted name, and I'd prefer a self explanatory name for the "memory is set to zero" guarantee (for non C programmers).

I'm all in for a single `GC.malloc(size, *, atomic = false, clear = true)` method (and an new `__crystal_malloc3` fun with all three args). I mean, [`atomic` is just a flag in my immix GC](https://github.com/ysbaddaden/gc/blob/master/src/immix.c#L173-L179) and it never sets memory to zero because crystal always does.

**Comment by ysbaddaden** [comment]
Anyway, having a mean to say "allocate and set to zero" and "allocate and don't set to zero" is needed for Crystal to tell it's intent to the allocator's interface, since each allocator has different guarantees:

- C: never zeroes (must use `calloc`);
- Boehm GC: `GC_malloc` always zeroes, `GC_malloc_atomic` never;
- Immix GC: never zeroes (DIY) but I could add an arg to do it at the most opportune moment.

**Comment by HertzDevil** [comment]
If we bypass the CRT and use Win32's [`HeapAlloc`](https://learn.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc) then there is also an optional argument specifying whether the memory should be cleared or not. On Windows `LibC.malloc` is only used in `-Dgc_none` and in the `wmain` entry point though

**Comment by beta-ziliani** [comment]
> I'm all in for a single GC.malloc(size, *, atomic = false, clear = true) method (and an new __crystal_malloc3(size, atomic, clear) fun)

We can't really make it just one function, because that will break backwards compatibility. But yes, we can have this generic version and keep the "atomic" version deprecated.

**Comment by ysbaddaden** [comment]
@BlobCodes oh, `allockind` looks very nice :+1:

**Comment by ysbaddaden** [comment]
I'm investigating memory allocations in codegen, and there are only a few calls:

- `Codegen#malloc` is called by `Codegen#malloc_closure` and `Codegen#malloc_aggregate`.
- `Codegen#malloc_atomic` is only called from `Codegen#allocate_aggregate`.
- `Codegen#array_malloc` and `Codegen#array_malloc_atomic` are only used to implement `Pointer.malloc`.

Of these use cases:

- `Codegen#malloc_closure`: I don't know if it needs the memory to be initialized to zero (it currently does or doesn't because of #14678);
- `Codegen#allocate_aggregate`: it needs the memory to be initialized to zero (and may do it twice => :bug:).
- `Pointer(T).malloc`: the memory must be initialized to zero (whatever if `T` has inner pointers).

So I don't think we have a need to distinguish a "please zero memory" vs "don't zero memory". Having the four `[array_]malloc[_atomic]` methods always return initialized memory feels like it would be enough. Then a check in `#malloc_aggregate` to avoid zeroing twice.

**Comment by ysbaddaden** [comment]
Urgh, sorry. I mixed up the issues. I'll move my comments _there_.

Note: I fail to see the older stdlib problem: a 1.12 or older compiler using the stdlib from 1.13+ (for example to compile Crystal 1.13 itself) will use the `__crystal_malloc64` fun and `GC.malloc` def from the 1.13+ stdlib that would now zero the memory. The worst to happen is the same as introducing new funs: the first time we compile Crystal 1.13 using Crystal 1.12, the compiler itself will keep zeroing the memory twice (but not the programs it compiles), and the compiler should recompile itself.

It would be an issue if using a 1.13 compiler with an older stdlib, but I would be very surprised if we supported that. We'd have the same issue with new funs that would end up calling the fallback LibC functions instead of GC methods :fearful: Definitely not supported.

**Comment by straight-shoota** [comment]
> It would be an issue if using a 1.13 compiler with an older stdlib, but I would be very surprised if we supported that. We'd have the same issue with new funs that would end up calling the fallback LibC functions instead of GC methods :fearful: Definitely not supported.

Yes, this is indeed a relevant use case. I suppose it's mostly for compiler/stdlib development. You can go back and build old versions with the current compiler. Of course it's not super important to ensure production-grade stability. But still worth considering.
Currently we can go back and build Crystal 1.8 with a 1.12 compiler. Older versions fail due to incorrectly mangled intrinsics after https://github.com/crystal-lang/crystal/pull/13164

**Comment by straight-shoota** [comment]
Summarizing the discussion here and in #14677 plus related, I believe we are agreed on the general design:

1. The GC interface should expose different mechanisms for allocating zeroed or non-zeroed memory. The allocator should be in charge of that because it can make the best decisions about allocated memory.
2. The stdlib and compiler can use the appropriate allocation strategy for efficiency and safety.
3. The compiler should select allocation strategy by required semantic contract, not by backend assumptions.
4. The C lib interface (used for codegen) needs separate functions for possible future optimizations, not a single function with flags.

I believe the biggest challenge to a clean API is the different semantics of `GC.malloc` and `GC.malloc_atomic`.
By following bdwgc we ended up with `GC.malloc` always zeroing memory, which is different from libc `malloc`, and made that an explicit property. But `GC.malloc_atomic` does not zero memory and is thus more similar to libc.
This reflects the most common use cases for each, but it's surprising.
Aligning their behaviour could cause friction. It might be easier to either live with the inconsistency, or introduce a completely new set of methods with a fresh naming scheme (e.g. `malloc{,_atomic}_{zeroed,dirty}`).

Another issue which we haven't explicitly talked about: The main premise of non-zeroing is that the user guarantees that any memory is written to before reading it. But that hardly holds when the GC could scan the memory before it's fully initialized. It might recognize dirty bytes as pointers.
So when the GC is involved, it seems that all non-atomic memory must be zeroed.
That reduces the practical use for an allocation function for scanned, non-zeroed memory to edge cases without garbage collection.
The compiler could still use such a function (if even makes sense semantically, see https://github.com/crystal-lang/crystal/issues/14687#issuecomment-2166280287), but with normal GC it would take the same code path as zeroed memory anyway.

**Comment by BlobCodes** [comment]
> the GC could scan the memory before it's fully initialized. It might recognize dirty bytes as pointers

I don't think this should be too much of a concern as this already happens on the stack. When returning from a method, the associated stack frame is also not cleared, and any variables previously used in the returning function could prevent the GC from freeing heap memory. With large `StaticArray`s used as buffers, this becomes even worse.

**Comment by ysbaddaden** [comment]
@straight-shoota The memory allocated by the GC as atomic won't be scanned by the GC, by definition. The memory can contain pointers, it will never be considered a live reference.

Anyway, not an issue in practice: at worst it would delay to next GC collection.

**Comment by straight-shoota** [comment]
@ysbaddaden The comment about GC + zeroing is explicitly about non-atomic (i.e. scanned) allocations.
For atomic allocations this doesn't matter of course, which is why `gc_malloc_atomic` returns non-zeroed memory.

@BlobCodes True, but the risk could be much higher on the heap.
The stack is only a couple of megabytes big.
A long living, multi gigabyte heap allocation has much more potential to cause unintended memory retention.

Of course, with garbage collection there is always a risk of interpreting actual data as pointers. But for uninitialized memory we can avoid that by zeroing.

**Comment by ysbaddaden** [comment]
Sorry I mixed paragraphs and misread.

No the risk is identical on the HEAP. It's entirely in the GC anyway and out of control.

**Comment by ysbaddaden** [comment]
But for how long? The compiler can only decide based on short term analysis, like the initialization of a struct or class, maybe a `StaticArray` when every entry is initialized. If there's any possibility for the memory to not be fully initialized, it shouldn't make a dirty malloc.

Hence: it's a low risk, it might delay a collectible object to the next GC collections, a few at worst.

The user making a dirty malloc and having leaks is out of scope (you asked for it, you deal with it).
